lib/Runtime/Library/ScriptFunction.cpp

```diff
@@ -83,50 +83,47 @@ namespace Js
+
         if (functionProxy->IsFunctionBody() && functionProxy->GetFunctionBody()->GetInlineCachesOnFunctionObject())
         {
-            Js::FunctionBody * functionBody = functionProxy->GetFunctionBody();
-            ScriptFunctionWithInlineCache* pfuncScriptWithInlineCache = scriptContext->GetLibrary()->CreateScriptFunctionWithInlineCache(functionProxy);
-            pfuncScriptWithInlineCache->SetEnvironment(environment);
-            JS_ETW(EventWriteJSCRIPT_RECYCLER_ALLOCATE_FUNCTION(pfuncScriptWithInlineCache, EtwTrace::GetFunctionId(functionProxy)));
-
-            Assert(functionBody->GetInlineCacheCount() + functionBody->GetIsInstInlineCacheCount());
-
+            FunctionBody * functionBody = functionProxy->GetFunctionBody();
             if (functionBody->GetIsFirstFunctionObject())
             {
-                // point the inline caches of the first function object to those on the function body.
-                pfuncScriptWithInlineCache->SetInlineCachesFromFunctionBody();
                 functionBody->SetIsNotFirstFunctionObject();
             }
             else
             {
+                ScriptFunctionWithInlineCache* pfuncScriptWithInlineCache = scriptContext->GetLibrary()->CreateScriptFunctionWithInlineCache(functionProxy);
                 // allocate inline cache for this function object
                 pfuncScriptWithInlineCache->CreateInlineCache();
-            }
 
-            pfuncScriptWithInlineCache->SetHasSuperReference(hasSuperReference);
+                Assert(functionBody->GetInlineCacheCount() + functionBody->GetIsInstInlineCacheCount());
+                if (PHASE_TRACE1(Js::ScriptFunctionWithInlineCachePhase))
+                {
+                    char16 debugStringBuffer[MAX_FUNCTION_BODY_DEBUG_STRING_SIZE];
 
-            if (PHASE_TRACE1(Js::ScriptFunctionWithInlineCachePhase))
-            {
-                char16 debugStringBuffer[MAX_FUNCTION_BODY_DEBUG_STRING_SIZE];
+                    Output::Print(_u("Function object with inline cache: function number: (%s)\tfunction name: %s\n"),
+                        functionBody->GetDebugNumberSet(debugStringBuffer), functionBody->GetDisplayName());
+                    Output::Flush();
+                }
 
-                Output::Print(_u("Function object with inline cache: function number: (%s)\tfunction name: %s\n"),
-                    functionBody->GetDebugNumberSet(debugStringBuffer), functionBody->GetDisplayName());
-                Output::Flush();
+                pfuncScript = pfuncScriptWithInlineCache;
             }
-            return pfuncScriptWithInlineCache;
         }
-        else
+
+        if (pfuncScript == nullptr)
         {
-            ScriptFunction* pfuncScript = scriptContext->GetLibrary()->CreateScriptFunction(functionProxy);
-            pfuncScript->SetEnvironment(environment);
+            pfuncScript = scriptContext->GetLibrary()->CreateScriptFunction(functionProxy);
+        }
 
```

Comment:
This is the part of the change that requires additional logic for release/1.9+. We can synch up to manage that.